validator_test.go

```diff
@@ -37,6 +37,10 @@ func TestValidXML(t *testing.T) {
 func TestColonsInLocalNames(t *testing.T) {
 	var err error
 
+	if el := tokenize(t, `<x::Root/>`).(xml.StartElement); el.Name.Local == `x::Root` {
```

Comment:
Let's use build tags to move all the tests that we are skipping to separate files. That avoids having to check for any custom condition and keeps things tidy.

Reply:
That was actually my original plan but I ended up with a bunch of if statements instead. The rationale being that with those we're able to do feature detection against the stdlib instead of just going by version: if the Go stdlib changes again in a future version, the feature detection-based approach means we don't end up skipping tests when we shouldn't.

Not sure if that makes sense though, considering build tags are much cleaner.

Reply:
I don't think we would be skipping any tests - either they fail, or they would pass. If the functionality breaks again, at most we add another build tag.

My proposal was to split the tests into files with `// +build !go1.17` for <1.17 and another with `// +build go1.17` for >= 1.17.

Reply:
We wouldn't have to literally skip any of the tests, but since we're testing against behaviors that only exist in Go ≤ 1.16, the test cases for 1.17+ would not cover those. XML parsing in 1.17 is still not anywhere near to being compliant with any spec, so it may well change again. If it does, it would be important to know that the validator still works as expected.

If we split this by build tags and just use the existing tests in the simplest way possible, there are two options: either the tests for 1.17+ just blindly succeed without actually testing anything, or they run the same validator calls as the tests for 1.16, but we change the expected output -- where we currently expect errors, we'd expect no errors in 1.17+. Even in the latter case, the failure mode would be towards the validator failing OPEN when in the case of a security library we want it to fail CLOSED.

Example:

Lets assume we fix the tests by splitting them and changing the expected output in 1.17+. In Go 1.16, a name like `x::Root` is parsed as `Space="x", Local=":Root"`. In Go 1.17 it's parsed as `Space="", Local="x::Root"`. Now maybe for Go 1.19 the Go team decides the new behavior is causing too many problems and that reverting to the old behavior is necessary; instead they fix the round-trip issue by making the encoder's behavior match that. If that happens, we'd *want* the tests to start failing again, because that would let us know that there has been a significant change, that the validator might need to account for.

Now, of course the new tests for 1.17+ could check that the tokens we use in the if statements in the current commit match what we expect and fail the test if they don't, but in that case we'd be literally testing the stdlib, which is generally not something you want to do. But maybe that makes sense in this case because this is a bit special anyway?

Thoughts?

Reply:
@agnivade I did the split by build tags and included checks for stdlib behavior in go1.17 tests. PTAL.

Reply:
Thank you for the detailed and illuminating explanation there @jupenur ! Really appreciated. I had to look up the difference between fail open vs fail closed :)

Just one clarification:

> Now, of course the new tests for 1.17+ could check that the tokens we use in the if statements in the current commit match what we expect and fail the test if they don't, but in that case we'd be literally testing the stdlib, which is generally not something you want to do. 

In your original commit, we were kinda already doing that isn't it? We were tokenizing and then skipping the test if it matched certain condition - indirectly locking the behavior. This seems like a more direct way of doing that.

But I totally agree with you on the fail mode behavior here as I did not take that into consideration before. So 0/5 if you want to revert to the old implementation. Otherwise this looks good to me as well.

Really appreciate your eye for detail Juho.

Reply:
Thanks Agniva.

> we were kinda already doing that isn't it?

Yes. That's absolutely right -- like you said, it was just less direct. That's why agree the current approach is better: we're still doing the same thing, but being more explicit about it so that it's clear we're doing something unusual. Plus with the split it was easier to test the happy path in 1.17, which got us a slightly better coverage.

So merging this in its current form. Thanks for all the feedback, always a learning experience for me, too!